File: code/Java/Chapter8/8-2/BubbleSort.java
import java.util.Arrays;

public class BubbleSort {
    public static void main(String[] args) {
        int[] arr = {5, 8, 6, 3, 9, 2, 1, 7};
        bubbleSort(arr);
        System.out.println(Arrays.toString(arr));
    }

    /**
     * Bubble Sort (original)
     */
     public static int[] bubbleSort(int[] arr) {
         for (int i = 0; i < arr.length; i++) {
             for (int j = 0; j < arr.length - i - 1; j++) {
                 if (arr[j] > arr[j + 1]) {
                     int temp = arr[j];
                     arr[j] = arr[j + 1];
                     arr[j + 1] = temp;
                 }
             }
         }

         return arr;
     }

    /**
     * Bubble Sort (optimized v1.0)
     */
//    public static int[] bubbleSort(int[] arr) {
//        for (int i = 0; i < arr.length; i++) {
//            boolean swapped = false;
//            for (int j = 0; j < arr.length - i - 1; j++) {
//                if (arr[j] > arr[j + 1]) {
//                    int temp = arr[j];
//                    arr[j] = arr[j + 1];
//                    arr[j + 1] = temp;
//                    swapped = true;
//                }
//            }
//            if (!swapped) {
//                break;
//            }
//        }
//
//        return arr;
//    }

    /**
     * Bubble Sort (optimized v2.0)
     */
//    public static int[] bubbleSort(int[] arr) {
//        int right = arr.length - 1;
//        for (int i = 0; i < arr.length; i++) {
//            boolean swapped = false;
//            int last = 0;
//            for (int j = 0; j < right; j++) {
//                if (arr[j] > arr[j + 1]) {
//                    int temp = arr[j];
//                    arr[j] = arr[j + 1];
//                    arr[j + 1] = temp;
//                    swapped = true;
//                    last = j;
//                }
//            }
//            if (!swapped) {
//                break;
//            }
//            right = last;
//        }
//
//        return arr;
//    }
}

File: code/Python/Chapter8/8-2/bubble_sort.py
def bubble_sort(lst):
    """Bubble Sort (original v1.0)"""
    for i in range(len(lst) - 1):
        for j in range(len(lst) - 1 - i):
            if lst[j] > lst[j + 1]:
                lst[j], lst[j + 1] = lst[j + 1], lst[j]
    
    return lst

# def bubble_sort(lst):
#     """Bubble Sort (optimized v2.0)"""
#     for i in range(len(lst) - 1):
#         swapped = False
#         for j in range(len(lst) - 1 - i):
#             if lst[j] > lst[j + 1]:
#                 lst[j], lst[j + 1] = lst[j + 1], lst[j]
#                 swapped = True
#         if not swapped:
#             break
    
#     return lst

# def bubble_sort(lst):
#     """Bubble Sort (optimized v2.1)"""
#     right = len(lst) - 1
#     for i in range(len(lst) - 1):
#         swapped = False
#         last = 0
#         for j in range(right):
#             if lst[j] > lst[j + 1]:
#                 lst[j], lst[j + 1] = lst[j + 1], lst[j]
#                 swapped = True
#                 last = j
#         if not swapped:
#             break
#         right = last

#     return lst

lst = [5, 8, 6, 3, 9, 2, 1, 7]
bubble_sort(lst)
print(lst)